paddle/scripts/teamcity/pr_build.sh

```diff
@@ -0,0 +1,116 @@
+#!/bin/bash
+#
+# This build script is used for TeamCity CI, we need to 
+# set the follwoing parameters in the project configuration:
+#   PARAMETER            | EXAMPLE    |   COMMENTS
+# CTEST_OUTPUT_ON_FAILURE   1         // whether output failure message
+# CTEST_PARALLEL_LEVEL      5         // parallel level for running ctest
+# PADDLE_VERSION            0.11.0    // the latest tag
+# WITH_AVX                  ON        // wheter the instruction support AVX
+# WITH_GOLANG               OFF       // don't build the Golang code
+# WITH_GPU                  ON        // running on GPU
+# WITH_UBUNTU_MIRROR        ON        // use unbuntu mirro to speed up dev Docker image building
+#
+
+set -xe
+
+PR_ID=$(echo %teamcity.build.vcs.branch.Paddle_PaddlePaddle% | sed 's/[^0-9]*//g')
+echo ${PR_ID}
+
+CHANGED_FILES=$(cat %system.teamcity.build.changedFiles.file% | awk -F : '{print $1}')
+if [ -z "$CHANGED_FILES" ]
+then
+  echo "empty changed files"
+else
+  echo "changed files: $CHANGED_FILES"
+  COUNT=$(echo $CHANGED_FILES | awk -F . '{print $NF}'|grep -i -E 'md|rst|jpg|png' | wc -l)
+  FILE_COUNT=$(echo $CHANGED_FILES | wc -l)
+  if [ "$COUNT" -eq "$FILE_COUNT" ]
+  then
+    echo "all readme files or pictures, skip CI"
+    exit 0
+  fi
+fi
+
+apt_mirror='s#http://archive.ubuntu.com/ubuntu#mirror://mirrors.ubuntu.com/mirrors.txt#g'
+
+export PADDLE_DEV_NAME=paddlepaddle/paddle:dev
+
+nvidia-docker run -i --rm -v $PWD:/paddle ${PADDLE_DEV_NAME} rm -rf /paddle/build
+
+# Use the cached thirdparty package if the MD5 is mathes.
+md5_content=$(cat \
+              ${PWD}/cmake/external/*.cmake \
+              ${PWD}/paddle/scripts/teamcity/pr_build.sh \
+              |md5sum | awk '{print $1}')
+
+tp_cache_dir=/root/.cache/third_party
+tp_cache_file=$cache_dir/ci_pr_${md5_content}.tar.gz
+
+if [ ! -d ${tp_cache_dir} ];then
+    mkdir -p ${tp_cache_dir}
+fi
+if [ -f ${tp_cache_file} ];then
+    mkdir -p ${PWD}/build
+    tar zxvf ${tp_cache_file}.tar.gz -C $PWD
+else
+    # clear the older tar files if MD5 has chanaged.
+    rm -rf $tp_cache_dir/*.tar.gz
+fi
+
+# Do not build dev image for now
+
+if [[ "%WITH_UBUNTU_MIRROR%" == "ON" ]]; then
```

Comment:
Can test if `Dockerfile` is not changed, then we can skip building the image.

Reply:
May not, a PR job may run on any node of the CI cluster, for one PR changes the `Dockerfile` and merged, the next PR run another node without changing of `Dockerfile`, and it would use the older dev Docker image, we do not want this. For another side,`docker build` would use the cached image Layer if `Dockerfile` has no changs, would not cost too much time.

Reply:
Does it make sense to tag the docker image with the hash of the Dockerfile? (and remove all other tags when a PR builds)
E.g.,
docker image:
paddlepaddle/paddle:dev-000000
paddlepaddle/paddle:dev-000001

If the hash is 000001, then use `paddlepaddle/paddle:dev-000001` and remove `paddlepaddle/paddle:dev-000000` to free up space.